File: frontend/types/slide.ts
export type TextElement = {
    type: "text";
    id: string;
    x: number;
    y: number;
    width?: number;
    height?: number;
    content: string
};

export type IconElement = {
    type: "icon";
    id: string;
    x: number;
    y: number;
    size: number;
    name: string;
};

export type SlideElement = TextElement | IconElement;

export type Slide = {
    id: string;
    elements: SlideElement[];
}

File: backend/app/models/slide.py
from pydantic import BaseModel, Field
from typing import Literal, Annotated

class TextElement(BaseModel):
    type: Literal["text"]
    id: str
    x: float
    y: float
    width: float | None = None
    height: float | None = None
    content: str

class IconElement(BaseModel):
    type: Literal["icon"]
    id: str
    x: float
    y: float
    size: float
    name: str

SlideElement = Annotated[
    TextElement | IconElement,
    Field(discriminator="type")
]

class Slide(BaseModel):
    id: str
    elements: list[SlideElement]